sql/wsrep_trans_observer.h

```diff
@@ -258,21 +258,32 @@ static inline bool wsrep_run_commit_hook(THD* thd, bool all)
 static inline int wsrep_before_prepare(THD* thd, bool all)
 {
   DBUG_ENTER("wsrep_before_prepare");
-  WSREP_DEBUG("wsrep_before_prepare: %d", wsrep_is_real(thd, all));
   int ret= 0;
   DBUG_ASSERT(wsrep_run_commit_hook(thd, all));
   if ((ret= thd->wsrep_parallel_slave_wait_for_prior_commit()))
   {
     DBUG_RETURN(ret);
   }
-  if ((ret= thd->wsrep_cs().before_prepare()) == 0)
+
+  ret= thd->wsrep_cs().before_prepare();
+
+  if (!ret)
   {
     DBUG_ASSERT(!thd->wsrep_trx().ws_meta().gtid().is_undefined());
+    /* Here we init xid with UUID and wsrep seqno. GTID is
+       set to undefined because commit order is decided later
+       in wsrep_before_commit(). wsrep_before_prepare() is
+       executed out of order. */
     wsrep_xid_init(&thd->wsrep_xid,
                    thd->wsrep_trx().ws_meta().gtid(),
-                   wsrep_gtid_server.gtid());
+                   wsrep_gtid_server.undefined());
   }
 
+  WSREP_DEBUG("wsrep_before_prepare: %d for %s ret=%s",
```

Comment:
Are these logging changes absolutely necessary? They don't seem to be related to the GTID issue, and there is a high risk that this extensive changes in this function create merge conflicts.

Reply:
I used these when debugging this issue as they output xid, but removed.